NIFI-8519 Adding HDFS support for NAR autoload

simonbence opened the pull request:
[NIFI-8519](https://issues.apache.org/jira/browse/NIFI-8519)

This is a proposal to extend NAR auto load capability with external sources. This PR contains the general concept to extend the behaviour and also provides a HDFS support. Other implementation might added simply by implementing the newly added interface. Thanks for taking your time with checking my PR!

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

ottobackwards commented:
I forked the nar 'system' a while ago, and re-implemented it on commons-vfs, with support for HDFS and a VFS classloader etc.

https://github.com/palindromicity/bundles

Just a different approach.  I look forward to see how this works out!

simonbence commented:
Thanks @markap14 ! I am on incorporating your ideas

ottobackwards commented:
One reason I implemented this by swapping out FileUtils to use VFS is that you don't need to change any of this, you just set the nifi properties paths that VFS urls and the rest just works.

markap14 commented:
@ottobackwards not sure VFS would provide what we need here. Not sure how it deals with all of the authentication etc. But most importantly, it doesn't really provide us the pluggability that we'd like to have here. We don't want to say "you can put extensions in your local file system or in HDFS." We want this to allow pluggability for any number of locations. We would also want to account for the fact that some systems may be offline, etc. and so we want to download the files locally, not just pull them on-demand each time. And we'd also have to consider all of the concerns with classloader isolation. The NAR Loader stuff at present is in the root lib. We certainly cannot put any sort of Hadoop libs, etc. at the root, so the logic needs to be isolated into NARs.

ottobackwards commented:
Absolutely,  I had different goals. 
Is there something written up on "We want this to allow pluggability for any number of locations. We would also want to account for the fact that some systems may be offline, ......"?

simonbence commented:
Hi @markap14 ! I incorporated your ideas, it really ended up better. The only exception is your last bullet point, which I extracted into a followup story https://issues.apache.org/jira/browse/NIFI-8602

markap14 reviewed (changes requested):
Thanks for the updates @simonbence! Things are looking really good here! I made a handful of minor comments inline. And I think we do need to make sure that we handle the @RequiresInstanceClassLoading annotation. I updated the code from your branch locally with the changes necessary for that and have done some testing. Everything is looking good. Will attach a patch for that to the Jira, as I think that's probably the easiest way to incorporate that in.

I did test using HDFSNarProvider to source extensions from the local file system. I also then configured a second provider to source from an S3 bucket and a third one to source from GCS. All worked great!

Most of the comments are around making sure we use UGI.doAs and class loaders properly. While I didn't run into any issues directly in my tests I do think it's possible for them to arise, depending on timing/what's been loaded, etc. So we need to make sure those calls are properly wrapped. Once that's good, I'll be a +1.

markap14 commented:
https://issues.apache.org/jira/secure/attachment/13025442/0001-NIFI-8519-Support-RequiresInstanceClassLoading-annot.patch is the patch that I attached.

markap14 commented:
One last thing that I think is necessary: please make sure that the admin guide is updated to describe the new properties.

simonbence commented on nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/nar/hadoop/HDFSNarProvider.java:
```diff
@@ -0,0 +1,231 @@
+            throw new IllegalArgumentException("At least one HDFS configuration resource is necessary");
+        }
+
+        this.sourceDirectory = new Path(Objects.requireNonNull(context.getParameters().get(SOURCE_DIRECTORY_PARAMETER)));
```
Path constructor provides some checks, but you are right with that it provides (and has) no information about what the path is for.

simonbence commented on nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/nar/hadoop/HDFSNarProvider.java:
```diff
@@ -0,0 +1,231 @@
+        final SocketFactory socketFactory = NetUtils.getDefaultSocketFactory(config);
+        Socket socket = null;
+
+        try {
```
As I see, because of the `"closeQuietly"` call in the `finally` block, it will not work. (The code is from `AbstractHadoopProcessor`)

simonbence commented on nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/nar/hadoop/HDFSNarProvider.java:
```diff
@@ -0,0 +1,231 @@
+        final Path path = getNarLocation(location);
+        final HdfsResources hdfsResources = getHdfsResources();
+
+        if (!hdfsResources.getFileSystem().exists(path)) {
```
I got this from `GetHDFS#performListing` which looks missing this, but now that I double check, it is wrapped in other places.

simonbence commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-nar-loading-utils/src/main/java/org/apache/nifi/nar/NarProviderTask.java:
```diff
@@ -0,0 +1,97 @@
+            try {
+                LOGGER.debug("Task starts fetching NARs from provider");
+                final Set<String> loadedNars = getLoadedNars();
+                final Collection<String> availableNars = narProvider.listNars();
```
This one actually breaks the behaviour. I am not sure why at this point, but it behaves the same way as if the context classloader would not be set.

markap14 commented:
Thanks @simonbence ! All looks good now. Fixed one checkstyle violation but all was good otherwise. +1 merged to main!
